marlowe-playground-client/src/MarloweEditor/BottomPanel.purs

```diff
@@ -6,21 +6,123 @@ import Prelude hiding (div)
+import MarloweEditor.Types (Action(..), BottomPanelView(..), MetadataAction(..), State, _editorErrors, _editorWarnings, _hasHoles, _metadataHintInfo, _showErrorDetail, contractHasErrors)
 import StaticAnalysis.BottomPanel (analysisResultPane, analyzeButton, clearButton)
 import StaticAnalysis.Types (_analysisExecutionState, _analysisState, isCloseAnalysisLoading, isNoneAsked, isReachabilityLoading, isStaticLoading)
 import Text.Parsing.StringParser.Basic (lines)
 
-panelContents :: forall p. State -> BottomPanelView -> HTML p Action
-panelContents state StaticAnalysisView =
+metadataList :: forall p. Map String String -> Set String -> (String -> String -> MetadataAction) -> (String -> MetadataAction) -> String -> String -> HTML p Action
+metadataList metadataMap hintSet setAction deleteAction typeNameTitle typeNameSmall =
+  div [ class_ $ ClassName "metadata-field-group" ]
+    if Map.isEmpty combinedMap then
+      []
+    else
+      [ h6_ [ em_ [ text $ typeNameTitle <> " descriptions" ] ]
+      , ol_
+          $ map
+              ( \(key /\ val) ->
+                  li_
+                    ( case val of
+                        Just (desc /\ needed) ->
+                          [ text $ typeNameTitle <> " " <> show key <> ": "
+                          , input
+                              [ type_ InputText
+                              , placeholder $ "Description for " <> typeNameSmall <> " " <> show key
+                              , class_ $ ClassName "metadata-input"
+                              , value desc
+                              , onValueChange $ Just <<< MetadataAction <<< setAction key
+                              ]
+                          , button
+                              [ classes [ if needed then plusBtn else minusBtn, smallBtn, ClassName "align-top" ]
+                              , onClick $ const $ Just $ MetadataAction $ deleteAction key
+                              ]
+                              [ text "-" ]
+                          ]
+                            <> if needed then [] else [ span [ classes [ ClassName "metadata-error", ClassName "metadata-not-used" ] ] [ text "Not used" ] ]
+                        Nothing ->
+                          [ span [ class_ (ClassName "metadata-error") ] [ text $ typeNameTitle <> " " <> show key <> " description not defined" ]
+                          , button
+                              [ classes [ plusBtn, smallBtn, ClassName "align-top" ]
+                              , onClick $ const $ Just $ MetadataAction $ setAction key mempty
+                              ]
+                              [ text "+" ]
+                          ]
+                    )
+              )
+          $ Map.toUnfoldable combinedMap
+      ]
+  where
+  mergeMaps :: (Maybe (String /\ Boolean)) -> (Maybe (String /\ Boolean)) -> (Maybe (String /\ Boolean))
```

Comment:
maybe add a type alias for this Tuple? or at least put a comment saying what it means?

Reply:
I think type alias here is overkill because it is just local to this function, but I will add a comment